client/bloodhound/client.go

```diff
@@ -0,0 +1,399 @@
+					s.log.Error(err, fmt.Sprintf("received GOAWAY from from AWS load balancer while requesting %s; attempt %d/%d; trying again", req.URL, currentAttept+1, s.maxRetries))
+					rest.VariableExponentialBackoff(s.retryDelay, currentAttept)
+					continue
+				}
+
+				// normal client error, dont attempt again
+				return nil, err
+			} else if res.StatusCode < http.StatusOK || res.StatusCode >= http.StatusBadRequest {
+				if res.StatusCode >= http.StatusInternalServerError {
+					// Internal server error, backoff and try again.
+					serverError := fmt.Errorf("received server error %d while requesting %v", res.StatusCode, req.URL)
+					s.log.Error(serverError, fmt.Sprintf("attempt %d/%d; trying again", currentAttept+1, s.maxRetries))
+
+					rest.VariableExponentialBackoff(s.retryDelay, currentAttept)
+					continue
+				}
+
+				// bad request we do not need to currentAttept
+				var body json.RawMessage
+				defer res.Body.Close()
+
+				if err := json.NewDecoder(res.Body).Decode(&body); err != nil {
+					return nil, fmt.Errorf("received unexpected response code from %v: %s; failure reading response body", req.URL, res.Status)
+				} else {
+					return nil, fmt.Errorf("received unexpected response code from %v: %s %s", req.URL, res.Status, body)
+				}
+			} else {
+				s.currentRequestCount += 1
+				if s.currentRequestCount >= s.requestLimit {
+					if err = s.resetConnection(); err != nil {
+						s.log.Error(err, "error resetting BHE http client connection")
+						return nil, err
+					}
+				}
+
+				return res, nil
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("unable to complete request to url=%s; attempts=%d;", req.URL, s.maxRetries)
+}
+
+// Ingest sends the ingest data to the BHE server and returns true if there were any errors while making the request
+func (s *BHEClient) Ingest(ctx context.Context, in <-chan []any) bool {
+	endpoint := s.bheUrl.ResolveReference(&url.URL{Path: "/api/v2/ingest"})
+
+	var (
+		hasErrors           = false
+		unrecoverableErrMsg = fmt.Sprintf("ending current ingest job due to unrecoverable error while requesting %v", endpoint)
+	)
+
+	for data := range pipeline.OrDone(ctx.Done(), in) {
+		var (
+			body bytes.Buffer
+			gw   = gzip.NewWriter(&body)
+		)
+
+		ingestData := models.IngestRequest{
+			Meta: models.Meta{
+				Type: "azure",
+			},
+			Data: data,
+		}
+
+		err := json.NewEncoder(gw).Encode(ingestData)
+		if err != nil {
+			s.log.Error(err, unrecoverableErrMsg)
+		}
+		gw.Close()
+
+		if req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint.String(), &body); err != nil {
+			s.log.Error(err, unrecoverableErrMsg)
+			return true
+		} else {
+			req.Header.Set("User-Agent", constants.UserAgent())
+			req.Header.Set("Accept", "application/json")
+			req.Header.Set("Content-Encoding", "gzip")
+
+			for currentAttempt := 0; currentAttempt <= s.maxRetries; currentAttempt++ {
```

Comment:
I renamed `retry` to `currentAttempt` as it made more sense
Since this logic is guaranteed to run at least one time before any retry can occur, it made more sense to me to rename this